unstructured/documents/elements.py

```diff
@@ -737,23 +749,27 @@ def convert_coordinates_to_new_system(
 
         return new_coordinates
 
-    @property
-    def id(self):
-        if self._element_id is None:
-            self._element_id = str(uuid.uuid4())
-        return self._element_id
-
-    def id_to_hash(self) -> str:
+    def id_to_hash(self, sequence_number: int) -> str:
         """Calculates and assigns a deterministic hash as an ID.
 
-        The hash ID is based on element's text.
+        The hash ID is based on element's text, sequence number on page,
+        page number and its filename.
 
-        Returns:
-            The first 32 characters of the SHA256 hash of the concatenated input parameters.
+        Args:
+            sequence_number: index on page
+
+        Returns: new ID value
         """
-        self._element_id = hashlib.sha256(self.text.encode()).hexdigest()[:32]
+        data = f"{self.metadata.filename}{self.text}{self.metadata.page_number}{sequence_number}"
```

Comment:
I think we're going to need some similar solution here, aren't we? `self.metadata.page_number` can be `None` still, right? And I think we said it should be 1 in that case, but I think it actually needs to be a little more sophisticated to deal with the "occasional None" case; otherwise page_number could go ... 6, 6, 1, 6, 7, 8, 1, ...

I'm inclined to think this method needs to take `page_number` as well as `sequence_number` (which should probably be named `page_sequence_number` btw).

Unless we've somehow established that all elements always have `.metadata.page_number` (and I don't think we have).

Reply:
are you sure we need to solve this?
we should not be writing `self.metadata.page_number` for this PR, only using it for hash calculations? and if it is None, just assume 1?

Reply:
@Crag: A couple things:
- This will be using "None" for the page-number, not "1".
- Therefore the "document-unique" promise is not guaranteed.

I had understood that the "document-unique" guarantee was a hard commit. But if we're satisfied with "usually mostly document-unique" then I suppose this will get it done :)

Reply:
if there are None page_number elements in a PDF, that sounds like a bug.
the biggest concern is PDF's. what case are you worried about?

Reply:
This code will be executed for all file-types, not just PDF, TIFF, and PPTX that "guarantee" page-numbers.

Reply:
So there are two cases across file-types:
- No page numbers
- Occasional omitted page numbers.

No-elements-have-page-numbers happens in formats like TXT and I'm sure several others. Occasional-omitted-page-numbers are like `PageBreak` and perhaps `PageNumber` elements are not assigned page-number metadata.

These are the cases that the proposed generator above addresses.

Reply:
> No-elements-have-page-numbers happens in formats like TXT
but then each element should have a different sequence number, as if it was one giant page

> Occasional-omitted-page-numbers are like PageBreak
if a PageBreak with no page number appears in a PDF or PPTX (or other "real page numbers" doc), imo that is a bug. the page number should be the page it is marking the end of.

Reply:
>> No-elements-have-page-numbers happens in formats like TXT

> but then each element should have a different sequence number, as if it was one giant page

Hmm, yes, good point. I suppose "None" is as good as "1" in that case, as long as we're consistent. I agree, that handles that case well enough since we aren't page-splitting those formats.

>> Occasional-omitted-page-numbers are like PageBreak

> if a PageBreak with no page number appears in a PDF or PPTX (or other "real page numbers" doc), imo that is a bug.

Okay, that's fair enough. We can check those formats and make sure they assign at least page-number metadata to those elements.